primitives/core/src/bandersnatch.rs

```diff
@@ -0,0 +1,1016 @@
+pub const SIGNING_CTX: &[u8] = b"SigningContext";
+
+// Max ring domain size.
+const RING_DOMAIN_SIZE: usize = 1024;
+
+#[cfg(feature = "full_crypto")]
+const SEED_SERIALIZED_LEN: usize = 32;
+
+// Short-Weierstrass form serialized sizes.
+const PUBLIC_SERIALIZED_LEN: usize = 33;
+const SIGNATURE_SERIALIZED_LEN: usize = 65;
+const PREOUT_SERIALIZED_LEN: usize = 33;
+const PEDERSEN_SIGNATURE_SERIALIZED_LEN: usize = 163;
+const RING_PROOF_SERIALIZED_LEN: usize = 592;
+
+// Max size of serialized ring-vrf context params.
+//
+// This size is dependent on the ring size.
+//
+// Some values:
+//  ring_size → ~serialized_size
+//   512        →  74 KB
+//  1024        → 147 KB
+//  2048        → 295 KB
+// NOTE: This is quite big but looks like there is an upcoming fix
+// in the backend.
+const RING_CONTEXT_SERIALIZED_LEN: usize = 147752;
+
+/// Bandersnatch public key.
+#[cfg_attr(feature = "full_crypto", derive(Hash))]
+#[derive(
+	Clone,
+	Copy,
+	PartialEq,
+	Eq,
+	PartialOrd,
+	Ord,
+	Encode,
+	Decode,
+	PassByInner,
+	MaxEncodedLen,
+	TypeInfo,
+)]
+pub struct Public(pub [u8; PUBLIC_SERIALIZED_LEN]);
+
+impl UncheckedFrom<[u8; PUBLIC_SERIALIZED_LEN]> for Public {
+	fn unchecked_from(raw: [u8; PUBLIC_SERIALIZED_LEN]) -> Self {
+		Public(raw)
+	}
+}
+
+impl AsRef<[u8; PUBLIC_SERIALIZED_LEN]> for Public {
+	fn as_ref(&self) -> &[u8; PUBLIC_SERIALIZED_LEN] {
+		&self.0
+	}
+}
+
+impl AsRef<[u8]> for Public {
+	fn as_ref(&self) -> &[u8] {
+		&self.0[..]
+	}
+}
+
+impl AsMut<[u8]> for Public {
+	fn as_mut(&mut self) -> &mut [u8] {
+		&mut self.0[..]
+	}
+}
+
+impl TryFrom<&[u8]> for Public {
+	type Error = ();
+
+	fn try_from(data: &[u8]) -> Result<Self, Self::Error> {
+		if data.len() != PUBLIC_SERIALIZED_LEN {
+			return Err(())
+		}
+		let mut r = [0u8; PUBLIC_SERIALIZED_LEN];
+		r.copy_from_slice(data);
+		Ok(Self::unchecked_from(r))
+	}
+}
+
+impl ByteArray for Public {
+	const LEN: usize = PUBLIC_SERIALIZED_LEN;
+}
+
+impl TraitPublic for Public {}
+
+impl CryptoType for Public {
+	#[cfg(feature = "full_crypto")]
+	type Pair = Pair;
+}
+
+impl Derive for Public {}
+
+impl sp_std::fmt::Debug for Public {
+	#[cfg(feature = "std")]
+	fn fmt(&self, f: &mut sp_std::fmt::Formatter) -> sp_std::fmt::Result {
+		let s = self.to_ss58check();
+		write!(f, "{} ({}...)", crate::hexdisplay::HexDisplay::from(&self.as_ref()), &s[0..8])
+	}
+
+	#[cfg(not(feature = "std"))]
+	fn fmt(&self, _: &mut sp_std::fmt::Formatter) -> sp_std::fmt::Result {
+		Ok(())
+	}
+}
+
+/// Bandersnatch signature.
+///
+/// The signature is created via the [`VrfSecret::vrf_sign`] using [`SIGNING_CTX`] as `label`.
+#[cfg_attr(feature = "full_crypto", derive(Hash))]
+#[derive(Clone, Copy, PartialEq, Eq, Encode, Decode, PassByInner, MaxEncodedLen, TypeInfo)]
+pub struct Signature([u8; SIGNATURE_SERIALIZED_LEN]);
+
+impl UncheckedFrom<[u8; SIGNATURE_SERIALIZED_LEN]> for Signature {
+	fn unchecked_from(raw: [u8; SIGNATURE_SERIALIZED_LEN]) -> Self {
+		Signature(raw)
+	}
+}
```

Comment:
Hmm... maybe a silly question, but why isn't this just simply `From`?

To me `UncheckedFrom` implies that there should be a `TryFrom` which does some extra invariant checking, but the `TryFrom` implementation doesn't check any extra invariants besides checking the size, which for a `From<[u8; SIGNATURE_SERIALIZED_LEN]>` would be checked at compile time.

Are there plans to extend the `TryFrom` so that it's actually checking something else besides that the size matches?

Reply:
The crypto primitives implements the same traits implemented by the other crypt types and using the same strategies.
Nevertheless, you are right and indeed my plan is to give some full overhaul to the crypto traits and interfaces in another PR